### DISCUSSION TITLE
Feature Request: allow to upgrade VM specs in a workflow/job basis

### DISCUSSION THREAD

**Discussion by stefanmaric** [community/community#26450] [Actions]
<p>Currently, for Linux and Windows runners, an Azure Standard_DS2_v2 VM is used, as stated here: <a href="https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners" rel="nofollow noopener">https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners</a></p>
<p>I haven’t run proper benchmarks, but we recently migrated some build tasks from CircleCI and Netlify to Github Actions, and I can say for sure the performance is subpar.</p>
<p>Most tasks take twice the time to complete in GA vs the free-tier counterparts on both, CircleCI and Netlify.</p>
<p>We have spent quite some time profiling and cutting seconds with workarounds, like for example, not using the <span class="mention">@actions</span>/use-node at all and using the nodejs version shipped with the runner — which is not ideal — to shave ~5 secs off jobs that require it. So it is not like we have done wrong translations while moving providers.</p>
<p>I’m excluding the fact GA lacks incremental cache: we have built our own persisted cache dirs using AWS S3, which of course has some overhead compared to having <span class="mention">@actions</span>/cache take care of it, as it would be the case in other CI providers. See: <a href="https://github.com/actions/cache/issues/109" rel="nofollow noopener">https://github.com/actions/cache/issues/109</a></p>
<p>We hace reached the point where we cannot optimize the workflows further, and it is still takes around 50% longer.</p>
<p>The only way to speed up our pipeline would be scaling the hardware vertically, but for that, we would need to use a self-hosted runner, which is definitely a big cost that makes GA economically inviable.</p>
<p>A feature to allow the selection of the Azure vm type in a workflow or job basis would solve this issue. Of course these would consume more credits per minute. CircleCI for example: <a href="https://circleci.com/docs/2.0/executor-types/" rel="nofollow noopener">https://circleci.com/docs/2.0/executor-types/</a></p>
<p>Related threads:</p>
<p><a href="https://github.community/t/standard-ds2-v2-vcpu-parallel-multithread-performance/17178">https://github.community/t5/GitHub-Actions/Standard-DS2-v2-vcpu-parallel-multithread-performance/m-p/41069</a></p>
<p><a href="https://github.community/t/insight-into-vm-performance/16199">https://github.community/t5/GitHub-Actions/Insight-into-VM-performance/m-p/31021</a><br>
<a href="https://github.community/t/higher-cpu-count-virtual-environments-planned/16338">https://github.community/t5/GitHub-Actions/Higher-CPU-count-virtual-environments-planned/m-p/32569</a></p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/stefanmaric">@stefanmaric</a> ,</p>
<p>Much appreciate your idea!</p>
<p>We can select different VMs for each job in a workflow(<a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idruns-on" rel="nofollow noopener">link</a>), but indeed it’s a limitation that cannot upgrade VM specs in the workflow/job basis.</p>
<p>According to the <a href="https://help.github.com/en/actions/getting-started-with-github-actions/about-github-actions#contacting-support" rel="nofollow noopener">policy</a>, it’s more appropriate to raise a feature_request ticket <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a> where github product manager will take a review. </p>
<p>Thanks.</p>
